odl/trafos/wavelet.py

```diff
@@ -834,14 +957,511 @@ def adjoint(self):
+                                           filterlength,
+                                           self.nscales, x.ctypes.data)
+            return self.range.element(x)
+
+        elif len(self.range.grid.shape) == 2:
+            (nx, ny) = (self.range.grid.shape[0], self.range.grid.shape[1])
+            coeff_in = coeff.copy()
+            coeff_in = coeff_in.asarray()
+            x = np.asarray(self.range.element())
+            bwt.adjointwavelet_transform2D(coeff_in.ctypes.data, nx, ny,
+                                           filterlength,
+                                           self.nscales, x.ctypes.data)
+            return self.range.element(x)
+
+        elif len(self.range.grid.shape) == 3:
+            (nx, ny, nz) = (self.range.grid.shape[0], self.range.grid.shape[1],
+                            self.range.grid.shape[2])
+            coeff_in = coeff.copy()
+            coeff_in = coeff_in.asarray()
+            x = np.asarray(self.range.element())
+            bwt.adjointwavelet_transform3D(coeff_in.ctypes.data, nx, ny, nz,
+                                           filterlength, self.nscales,
+                                           x.ctypes.data)
+            return self.range.element(x)
+
+    @property
+    def adjoint(self):
+        """The biorthogonal wavelet transform."""
+        return BiorthWaveletTransform(dom=self.range, nscales=self.nscales,
+                                      wbasis=self.wbasis)
+
+    @property
+    def inverse(self):
+        """The inverse of the adjoint of biorthogonal wavelet transform."""
+        return InverseAdjBiorthWaveletTransform(ran=self.range,
+                                                nscales=self.nscales,
+                                                wbasis=self.wbasis)
+
+
+class InverseBiorthWaveletTransform(Operator):
+    """
+    """
+
+    def __init__(self, ran, nscales, wbasis):
+        """Initialize a new instance.
+
+        Parameters
+        ----------
+        ran : `DiscreteLpVector` ("image domain")
+
+        nscales : `int`
+            Number of scaling levels in the wavelet transform.
+            The maximum number of usable scales with J.-O.S.
+            biorthogonal wavelets can be determined using xxx
+        wbasis : `str`
+            (``'JOSbiorthN'``) J.-O. Stromberg's biorthogonal wavelets
+            ``N`` is the filter length.
+            Possible values are:
+
+            1, 3, 5, 7 or 9
+
+        Signal extention mode: reflection
+        """
+        self.nscales = int(nscales)
+        self.wbasis = wbasis
+
+        if not isinstance(ran, DiscreteLp):
+            raise TypeError('domain {!r} is not a `DiscreteLp` instance.'
+                            ''.format(ran))
+
+        if ran.exponent != 2.0:
+            raise ValueError('domain Lp exponent is {} instead of 2.0.'
+                             ''.format(ran.exponent))
+
+        ndim = ran.ndim
```

Comment:
This piece of code (following 10 lines) is written many times. Perhaps make a utility?